From 603edcbff5301cc26f6f46ed953d8e0b342bd9d4 Mon Sep 17 00:00:00 2001
From: Akira Ajisaka <aajisaka@apache.org>
Date: Tue, 26 Sep 2017 01:47:02 +0900
Subject: [PATCH 2684/2748] HADOOP-14892. MetricsSystemImpl should use
 Time.monotonicNow for measuring durations.
 Contributed by Chetna Chaudhari.

(cherry picked from commit 3a10367a17b99e2c3035ec54048f94f334d93227)
(cherry picked from commit 6004fb46761810ef87367c5207fafcdbb6507c39)
(cherry picked from commit ea7e655897e1084063b2a570dc59c0fc8fcc7ad4)

Change-Id: I472b8e8b964258a955febc73ce1b817ed5eb9a78
---
 .../hadoop/metrics2/impl/MetricsSystemImpl.java    |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
index 422f9cd..728faca 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java
@@ -417,10 +417,10 @@ public synchronized MetricsBuffer sampleMetrics() {
 
   private void snapshotMetrics(MetricsSourceAdapter sa,
                                MetricsBufferBuilder bufferBuilder) {
-    long startTime = Time.now();
+    long startTime = Time.monotonicNow();
     bufferBuilder.add(sa.name(), sa.getMetrics(collector, true));
     collector.clear();
-    snapshotStat.add(Time.now() - startTime);
+    snapshotStat.add(Time.monotonicNow() - startTime);
     LOG.debug("Snapshotted source "+ sa.name());
   }
 
@@ -433,7 +433,7 @@ private void snapshotMetrics(MetricsSourceAdapter sa,
   synchronized void publishMetrics(MetricsBuffer buffer, boolean immediate) {
     int dropped = 0;
     for (MetricsSinkAdapter sa : sinks.values()) {
-      long startTime = Time.now();
+      long startTime = Time.monotonicNow();
       boolean result;
       if (immediate) {
         result = sa.putMetricsImmediate(buffer); 
@@ -441,7 +441,7 @@ synchronized void publishMetrics(MetricsBuffer buffer, boolean immediate) {
         result = sa.putMetrics(buffer, logicalTime);
       }
       dropped += result ? 0 : 1;
-      publishStat.add(Time.now() - startTime);
+      publishStat.add(Time.monotonicNow() - startTime);
     }
     droppedPubAll.incr(dropped);
   }
-- 
1.7.9.5

